### PR TITLE
fix: implement Merkle proof verification for legacy funding txs

### PR DESCRIPTION
**Pull request by piyush-1337**
# Pull Request

## Description

Resolves the `TODO` in `legacy_swap.rs` regarding missing Merkle proof validation.

Prior to this change, the Taker was populating the `funding_tx_merkleproof` field with an empty string, and the Maker was accepting it without cryptographically verifying the funding transaction's existence on the blockchain.

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [x] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [ ] Neither (infrastructure / tooling only)

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [x] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [x] I ran `cargo +nightly fmt --all` and committed the result
- [x] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [x] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [x] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [x] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [x] All unit tests pass (`cargo test`)
- [x] Integration tests pass (`cargo test --features integration-test`)
- [ ] Changes were manually tested on **regtest**
- [x] End-to-end maker ↔ taker swap tested (where applicable)
- [x] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [x] Complex logic is commented

### Security & Privacy (Critical)
- [x] This change preserves **trustlessness** and **atomic swap guarantees**
- [x] No regression in **sybil resistance** or **fidelity bond** logic
- [ ] Tor anonymity and P2P message flow were reviewed
- [x] No new attack vectors or trust assumptions introduced
- [x] Edge cases and error handling considered
- [x] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [x] `integration-test` feature flag is not reachable in production code paths

## How to Test

```bash
CFLAGS="-Wno-int-conversion" cargo test --features integration-test
```

## Summary by CodeRabbit

* **New Features**
  * On-chain merkle-proof verification for submitted funding transactions: proofs are fetched from the connected wallet and verified before swaps proceed.
  * Verification now retries for a short period with pauses to allow proofs to appear on-chain.
  * Wallet-backed verification enforces strict matching of the proved transaction.

* **Bug Fixes**
  * Proof validation fails fast on missing, empty, or mismatched proofs, preventing unverified funding from progressing.
  * Retrieval errors during proof collection now propagate immediately, eliminating placeholder proofs.

### CODE REVIEW

**Comment by piyush-1337** [comment]
@coderabbitai review

**Comment by piyush-1337** [comment]
@coderabbitai review

**Comment by mojoX911** [comment]
@piyush-1337 initial change set looks good. But the same logic also applies for Taproot protocol too. So can be replayed there. Not only for legacy.

File: src/maker/api.rs [hunk 1]
```diff
@@ -1202,6 +1257,11 @@ impl MakerTrait for MakerServer {
                 .ok_or(MakerError::General("Funding output not found"))?
                 as u32;
 
+            let funding_txid = funding_info.funding_tx.compute_txid();
+
+            // Verify the taker-provided SPV proof commits to this funding transaction.
+            self.wait_and_verify_tx_out_proof(&funding_txid, &funding_info.funding_tx_merkleproof)?;
```
Comment (mojoX911) [inline src/maker/api.rs]: Put this verification after confirmation is checked below. So then you don't need to wait for anything. If the tx is confirmed, merkle_proof will always exist.

File: src/taker/legacy_swap.rs [hunk 2]
```diff
@@ -871,25 +873,44 @@ impl Taker {
...
+                        &hashlock_nonce,
+                    )| {
+                        let txids = [funding_tx.compute_txid()];
+
+                        // Try to get blockhash to avoid requiring -txindex
+                        let blockhash = wallet
+                            .rpc
+                            .get_transaction(&txids[0], None)
+                            .ok()
+                            .and_then(|info| info.info.blockhash);
```
Comment (mojoX911) [inline src/taker/legacy_swap.rs]: Txindex is already a requirement for backend nodes for other stuffs. So we are not saving anything by not requiring it here. 

We rather can save one extra rpc call. RPC calls are costly.

File: src/wallet/rpc.rs [hunk 3]
```diff
@@ -237,4 +237,37 @@ impl Wallet {
...
+    pub fn verify_tx_out_proof(
+        &self,
+        expected_txid: &bitcoin::Txid,
+        proof_hex: &str,
+    ) -> Result<(), WalletError> {
+        if proof_hex.is_empty() {
+            return Err(WalletError::General(
+                "Missing funding tx merkle proof".to_string(),
+            ));
+        }
```
Comment (mojoX911) [inline src/wallet/rpc.rs]: If we are always verifying the proof after confirmation, in no situation proof_hex can be empty here. So we can safely assume here.

File: src/wallet/rpc.rs [hunk 4]
```diff
@@ -237,4 +237,37 @@ impl Wallet {
...
+
+        let proof_txids: Vec<bitcoin::Txid> = self
+            .rpc
+            .call("verifytxoutproof", &[json!(proof_hex)])
+            .map_err(WalletError::Rpc)?;
+
+        if proof_txids.is_empty() {
+            return Err(WalletError::General(
+                "proof failed SPV verification or not in best chain".to_string(),
+            ));
```
Comment (mojoX911) [inline src/wallet/rpc.rs]: Add a special error variant for SPV verification failure instead of general string.

File: src/wallet/rpc.rs [hunk 5]
```diff
@@ -237,4 +237,37 @@ impl Wallet {
...
+        }
+
+        if proof_txids.len() != 1 || proof_txids.first() != Some(expected_txid) {
+            return Err(WalletError::General(format!(
+                "Funding tx merkle proof does not match expected txid. Expected: {}, Got: {:?}",
+                expected_txid, proof_txids
+            )));
+        }
```
Comment (mojoX911) [inline src/wallet/rpc.rs]: You can compress this whole verification step into one statement by doing `proof_txids == expected_vec`. No need for empty or single input checks. All of that will can be implied in just one vector match check.

Use one SPV error variant for all failures. `MerkleProofInvalid | expected: {} | got: {}`. 

This whole function then could just inline in the call site with one or two statements at max instead of having separate routine for it.

File: src/wallet/rpc.rs (hunk 4 above)
Comment (mojoX911) [reply src/wallet/rpc.rs]: just a `WalletError::MerkleProofInvalid` should be enough to cover everything.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Few more structural suggestions.

**Comment by piyush-1337** [comment]
@mojoX911  For Taproot, what exact equivalent check do you want? There is no ProofOfFunding / funding_tx_merkleproof message in the Taproot flow. Right now the maker checks that the incoming contract tx is visible on-chain, while the taker waits for confirmations before sending TaprootContractData. Do you want a new proof field/message added to the Taproot protocol too?

**Comment by mojoX911** [comment]
> @mojoX911 For Taproot, what exact equivalent check do you want? There is no ProofOfFunding /

Hey @piyush-1337 . Nothing is needed for this. The PR is good enough at this stage. Although it's a bit redundant. We are already waiting for confirmations, so a Merkle proof isn't necessary.

Will give it one more through pass. If it looks good, we can include it for the legacy swaps for completeness.

**Review by mojoX911 — APPROVED** [review]
Ack
